head.go

```diff
@@ -192,11 +210,59 @@ func (h *HeadBlock) Meta() BlockMeta {
 	return m
 }
 
-func (h *HeadBlock) Dir() string         { return h.dir }
-func (h *HeadBlock) Persisted() bool     { return false }
-func (h *HeadBlock) Index() IndexReader  { return &headIndexReader{h} }
+// Dir implements headBlock
+func (h *HeadBlock) Dir() string { return h.dir }
+
+// Persisted implements headBlock
+func (h *HeadBlock) Persisted() bool { return false }
+
+// Index implements headBlock
+func (h *HeadBlock) Index() IndexReader { return &headIndexReader{h} }
+
+// Chunks implements headBlock
 func (h *HeadBlock) Chunks() ChunkReader { return &headChunkReader{h} }
 
+// Delete implements headBlock
+func (h *HeadBlock) Delete(mint int64, maxt int64, ms ...labels.Matcher) error {
+	h.mtx.RLock()
+
+	ir := h.Index()
+
+	pr := newPostingsReader(ir)
+	p, absent := pr.Select(ms...)
+
+	h.mtx.RUnlock()
+
+	h.mtx.Lock() // We are modifying the tombstones here.
+	defer h.mtx.Unlock()
```

Comment:
Switching from read to write lock is always a bit dangerous as stuff can happen in between. If we always have to the acquire the wlock anyway, it's typically not worth the headache. We can just acquire the write lock from the beginning.